### DISCUSSION TITLE
Why would this code not be breaking the interval at the end of the cycle?

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#64026] [Help] [answered]
### Summary

From the code below.

The code works fine with the API and sending emails properly, but it isn't breaking the intervals once the length of the document has been reached.

Also, I can see the first 2 console.logs (we are running and data), but once the `fetc`h function executes, it seems like the rest of the code isn't running except for the time for the` setInterval`  function..

The length of the document is only 2 objects, and I can see the each object being console.log properly each time separately for the first 2 runs.... then to code just breaks.. 

There are never console.logs for:
fetch function
date run time
arrayIndex, quotes.lenght
we are done 

And obviously it isn't catching this part of the code to break the interval:

```
if (arrayIndex === quotes.length) {
        clearInterval(myIntervals);
        arrayIndex = 0;
        console.log("We are done!");
      }

```
The full code is this:

```
import React from "react";
import quotes from "../thejsons/test-quote.json";

export default function ContactForm() {
  const handleSubmit = async () => {
    const d = new Date();
    const month = d.getMonth();
    const days = d.getDate();
    const year = d.getFullYear();
    const hour = d.getHours();
    const minutes = d.getMinutes();
    const seconds = d.getSeconds();

    var arrayIndex = 0;

    console.log(
      `We are moving on ${month}/${days}/${year} at ${hour}:${minutes}:${seconds}s`
    );

    var myIntervals = setInterval(async function () {
      const currentQuote = quotes[arrayIndex++];
      const data = {
        contact: currentQuote.Contact,
        email: currentQuote.Email,
        quote: currentQuote.quote,
        id: currentQuote.id,
      };

      console.log(data);

      await fetch("/api/quote-email", {
        method: "POST",
        headers: {
          Accept: "application/json, text/plain, */*",
          "Content-Type": "application/json",
        },
        body: JSON.stringify(data),
      }).then((res) => {
        console.log("Response received");

        if (res.status === 200) {
          console.log("Response succeeded!");
        }
      });

      const d = new Date();
      const month = d.getMonth();
      const days = d.getDate();
      const year = d.getFullYear();
      const hour = d.getHours();
      const minutes = d.getMinutes();
      const seconds = d.getSeconds();
      timesRun += 1;
      console.log(
        `we ran ${timesRun} times at ${month}/${days}/${year} at ${hour}:${minutes}:${seconds}s`
      );

      console.log(arrayIndex, quotes.length);

      if (arrayIndex === quotes.length) {
        clearInterval(myIntervals);
        arrayIndex = 0;
        console.log("We are done!");
      }
    }, 20 * 1000);
  };

  return (
    <div>
      <div className="col-md-12 main-contact-form-div">
        <h4 className="get-in-touch">Get in touch.</h4>
        <br />
        <br />
        <button className="btn btn-primary form-button" onClick={handleSubmit}>
          click me
        </button>
      </div>
    </div>
  );
}




```

**Comment by icyJoseph** [comment]
Hi,

The problem is that `timesRun` is not defined, so the handleSubmit function, crashes, and then it never gets to the clear interval part. 

Either define it at the top of the file, or just plain remove it (`timesRun`).

Also, please avoid using `var`, prefer `let` if you need to mutate stuff.

So you may have:

```ts
 let arrayIndex = 0;
 let timesRun = 0; // or just plain remove it's usage within the interval callback
    console.log(
      `We are moving on ${month}/${days}/${year} at ${hour}:${minutes}:${seconds}s`
    );

 let myIntervals = setInterval(async function () {
```

Last but not least, intervals are not 100% reliable, timers really, but I think in this case you are probably fine... unless the user tabs away?

**Reply by Elindo586** [reply]
Hey!

I just saw this post..  yes, the timesRun wasn't there, then I added to it, but that just counts the numbers of time running, and it doesn't do anything with the `clearInterval()` function.... let me see what happens if I take it out again.

**Reply by Elindo586** [reply]
So yes, the code breaks if the `timeRuns` is not there.. but I did fix that before trying again...

**Reply by icyJoseph** [reply]
Well at least in the code you shared here, timesRun is never declared, so under strict mode, the program crashes.

**Comment by Elindo586** [comment]
So, if I move everything that wasn't showing above my `fetch` function, then everything works, but WHY???

^^^  This is not really the ideal way, as it gives me information before the cycle is going to be done.

Interesting in plain REACT the `setIntervals` is working similar to the original code to communicate with a Tiwtter API...  

When I do a similar code using the `setIntervals`  to communicate with the email API, then I don't get the expected behavior.

**Comment by Elindo586** [comment]
So this is the new code... works on clearing the interval, but I had to move it above the `fetch` function...   WHY does it work above the fetch function and not below it?

```
import React from "react";
import quotes from "../thejsons/test-quote.json";

export default function ContactForm() {
  const handleSubmit = async () => {
    const d = new Date();
    const month = d.getMonth();
    const days = d.getDate();
    const year = d.getFullYear();
    const hour = d.getHours();
    const minutes = d.getMinutes();
    const seconds = d.getSeconds();

    var timesRun = 0;
    var arrayIndex = 0;

    console.log(
      `We are moving on ${month}/${days}/${year} at ${hour}:${minutes}:${seconds}s`
    );

    var myIntervals = setInterval(async function () {
      const currentQuote = quotes[arrayIndex++];
      const data = {
        contact: currentQuote.Contact,
        email: currentQuote.Email,
        quote: currentQuote.quote,
        id: currentQuote.id,
      };

      console.log(data);

      if (arrayIndex === quotes.length) {
        clearInterval(myIntervals);
        arrayIndex = 0;
        console.log("We are done!");
      }
      const d = new Date();
      const month = d.getMonth();
      const days = d.getDate();
      const year = d.getFullYear();
      const hour = d.getHours();
      const minutes = d.getMinutes();
      const seconds = d.getSeconds();
      timesRun += 1;
      console.log(
        `we ran ${timesRun} times at ${month}/${days}/${year} at ${hour}:${minutes}:${seconds}s`
      );

      console.log(arrayIndex, quotes.length);

      await fetch("/api/quote-email", {
        method: "POST",
        headers: {
          Accept: "application/json, text/plain, */*",
          "Content-Type": "application/json",
        },
        body: JSON.stringify(data),
      }).then((res) => {
        console.log("Response received");

        if (res.status === 200) {
          console.log("Response succeeded!");
        }
      });
    }, 10 * 1000);
  };

  return (
    <div>
      <div className="col-md-12 main-contact-form-div">
        <h4 className="get-in-touch">Get in touch.</h4>
        <br />
        <br />
        <button className="btn btn-primary form-button" onClick={handleSubmit}>
          click me
        </button>
      </div>
    </div>
  );
}

```

**Reply by icyJoseph** [reply]
And you see the response success logs too? I wonder if your API route is not responding at all. So, it'd work because this would be doing an awkward fire and forget pattern.

**Reply by Elindo586** [reply]
No, there is no response success.   from the front page I see the console.logs above the fetch function. I have to use the browser developer tools to see them.   On the api page I see the logs from the api on the VS code Terminal..... Now, the api page is sending emails and doing the work it should do on its side apparently.

**Reply by icyJoseph** [reply]
So the issue is most likely in the API call. Either it crashes, or doesn't respond to the client, causing perhaps a timeout, which would also crash the interval callback. 

Could you debug the API function to make sure it responds and doesn't error itself?

**Reply by icyJoseph** [reply]
You can also chain a .catch to the fetch promise and inspect the returned object.

**Reply by Elindo586** [reply]
This one didn't do anything:
```
await fetch("/api/quote-email", {
        method: "POST",
        headers: {
          Accept: "application/json, text/plain, */*",
          "Content-Type": "application/json",
        },
        body: JSON.stringify(data),
      })
        .then((res) => {
          console.log("Response received");

          if (res.status === 200) {
            console.log("Response succeeded!");
          }
        })
        .catch((error) => {
          console.log("Error", error);
        });
```

On the API side also didn't give me the last console.log that I just added.... maybe I need to give it more time between runs??? 

```
await new Promise((req, res) => {
    transporter.sendMail(mailData, function (err, info) {
      // if (err) console.log(err);
      // // else console.log(info);
    });
  });

  console.log("email sent");
```

The emails are going out properly, but  I don't see to get the feedback from the API..

**Reply by Elindo586** [reply]
I gave it 100 seconds between cycles, and still no console.logs from the fetch..

**Reply by Elindo586** [reply]
This is working from the API page...

```
await new Promise((req, res) => {
    transporter.sendMail(mailData, function (err, info) {
      if (err) console.log(err);
      else console.log("email sent");
    });
  });
```

The front page still gets no info from the API to display.

**Reply by icyJoseph** [reply]
Right, but you never resolve that promise.

In the if error branch you should call the reject function, which you have called res... And in the success branch you should call the resolve function, which you have called, req...

Something like this:

```ts
await new Promise((resolve, reject) => {
    transporter.sendMail(mailData, function (err, info) {
      if (err) { 
          console.log(err);
          reject(err);
       } else {
          console.log("email sent");
          resolve()
       }
    });
  });

```

And that'll allow the API to resume, and respond to the client.

**Reply by Elindo586** [reply]
Ah ha!.,

```
await new Promise((resolve, reject) => {
    transporter.sendMail(mailData, function (err, info) {
      if (err) { 
          console.log(err);
          reject(err);
       } else {
          console.log("email sent");
          resolve()
       }
    });
  });

```
Got this error:

`API resolved without sending a response for /api/quote-email, this may result in stalled requests.`

^^  Now great....   this is an api page for nodemailer using Outlook365..   Is there a way to solve it? Do I need to contact Microsoft?  

```
export default async function stuff(req, res) {
  let nodemailer = require("nodemailer");
  const transporter = nodemailer.createTransport({
    service: "Outlook365",
    auth: {
      user: process.env.EMAIL,
      pass: process.env.EPASSWORD,
    },
  });
```

**Reply by icyJoseph** [reply]
The issue now is that, your promise resolves, and moves to the next step, and then you should actually `res.end()` or something like that, to tell Next.js that you are done, and they can respond back to the client. Do you have a link to a repository where this lives?

**Comment by Elindo586** [comment]
Do you want to see the repo?

Here is the api page:
https://github.com/Elindo586/nextprojectd/blob/main/pages/api/quote-email.js''

Here is the frontend page:
https://github.com/Elindo586/nextprojectd/blob/main/pages/file.js

Where should the `res.end()` go?

**Reply by icyJoseph** [reply, marked as answer]
Here, at https://github.com/Elindo586/nextprojectd/blob/main/pages/api/quote-email.js#L59, add this:

```ts
   return  res.status(200).json({ message: 'Email sent!' })
```

**Reply by Elindo586** [reply]
It worked.

So I just needed to add the res.status(200) to get it back to the frontend... silly me..
